tasks/h8is2w8/02-graph-dp/src/break-text.lisp

```diff
@@ -0,0 +1,102 @@
+(in-package :break-text)
+
+(defvar *words-dict* (make-hash-table :test #'equal))
+(defvar *bi-grams-dict* (make-hash-table :test #'equal))
+
+(defun load-dict-from-file (file fn)
+  (with-open-file (in file)
+    (loop :for line = (read-line in nil)
+       :while line :do (funcall fn line))))
+
+(defun reset-dict (dict)
+  (setf dict (make-hash-table :test #'equal)))
+
+(defun add-word (str)
+  (setf (gethash (string-right-trim '(#\Return) str) *words-dict*) t))
+
+(defun get-word (str)
+  (gethash (string-downcase str) *words-dict*))
+
+(defun add-bi-gram (line)
+  (destructuring-bind (bi-gram count) (cl-ppcre::split "\\t" line)
+    (setf (gethash bi-gram *bi-grams-dict*)
+	  (parse-integer count))))
+
+(defun get-bi-gram (bi-gram)
+  (gethash bi-gram *bi-grams-dict* 0))
+
+(defun load-words-dict (&optional (file "data/dict_en.txt"))
+  (load-dict-from-file file #'add-word))
+
+(defun load-bi-grams-dict (&optional (file "data/short_bi_grams.txt"))
+  (load-dict-from-file file #'add-bi-gram))
+
+(defun work-setup ()
+  (load-words-dict)
+  (load-bi-grams-dict))
+
+(defun max-value-index (array)
+  "Find index of the last maximum value."
+  (loop
+     :with max-index = 0
+     :with max-value = (aref array 0)
+     :for x :across array
+     :for i :from 0
+     :when (>= x max-value)
+     :do (progn (setf max-index i) (setf max-value x))
+     :finally (return max-index)))
+
+(defun break-text (str)
+  (let* ((graph (text->graph str))
+	 (sorted (reverse (tsort graph)))
+	 (costs (make-array (nodes-count graph)))
```

Comment:
I'd call it not `costs` but `freqs` or `weights`, because you want to select the lowest cost but the highest weight